feat: EzRating component

danidewitt opened the pull request:
## What did we change?
- Added an icon rating component `EzRating`

## Why are we doing this?
- To be able to create a star rating component (flexible enough to use with other icons, too)

## Screenshot(s) / Gif(s):
<img width="1034" alt="Screen Shot 2022-02-15 at 1 24 07 PM" src="https://user-images.githubusercontent.com/3790037/154125153-c81ef9a4-b848-4eeb-a541-1a57565b74c5.png">
<img width="641" alt="Screen Shot 2022-02-15 at 1 24 24 PM" src="https://user-images.githubusercontent.com/3790037/154125163-88aa1fb5-c6bc-43ab-8520-44e77b744ccc.png">
<img width="654" alt="Screen Shot 2022-02-15 at 1 24 35 PM" src="https://user-images.githubusercontent.com/3790037/154125166-955c7f9f-f25a-4560-a535-b031403e5a74.png">

## Checklist

- [x] Provide test coverage for the changes made
- [x] Create a changeset (`npm run changeset`) with a summary of the changes

[Contributing guidelines](https://ezcater.github.io/recipe/guides/contributing/)

ez-alexfrazer commented on src/components/EzRating/Implementations/EzRatingCustom/EzRatingCustom.tsx:
```diff
@@ -0,0 +1,40 @@
+  const emptyIconCount = max - Math.ceil(valueRoundedToClosestHalfInteger);
+
+  return (
+    <span role="img" aria-label={label}>
```
🎱 would this be more accessible as a [meter](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meter) element? It also allows for some more granular control of [styling](https://css-tricks.com/html5-meter-element/)

morgannegagne commented on src/components/EzRating/EzRating.md:
```diff
@@ -0,0 +1,101 @@
+
+---
+
+## Examples
```
Is it worth having an example with a `label` prop?

ez-alexfrazer commented on src/components/EzRating/Implementations/EzRatingCustom/EzRatingCustom.tsx:
```diff
@@ -0,0 +1,40 @@
+  const emptyIconCount = max - Math.ceil(valueRoundedToClosestHalfInteger);
+
+  return (
+    <span role="img" aria-label={label}>
```
I think the accessibility on this might be unclear, would a [`meter`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meter) component make sense in this case?

> The <meter> [HTML](https://developer.mozilla.org/en-US/docs/Web/HTML) element represents either a scalar value within a known range or a fractional value.

danidewitt replied:
we're doing what [mui does for rating](https://mui.com/components/rating/#accessibility) - they use a `span` with `role="img"` and `aria-label=`${label}``

danidewitt commented on src/components/EzRating/EzRating.types.ts:
```diff
@@ -0,0 +1,17 @@
+  max: number;
+  value: number;
+  label: string;
+  emptyIcon: EzIconTypes;
+  halfIcon: EzIconTypes;
+  fullIcon: EzIconTypes;
```
I'm still thinking about this default icon bit, maybe we can just import the three free default icons and it be ok... I'll add it and we can test the performance of it.

danidewitt replied:
@noranda and @maggiedbaker i think this thread answers your questions

noranda reviewed (changes requested):
This looks great!

I like the idea of having a default icon (star?). Maybe we can use a free-license icon for this so we don't run into issues. We also might want to think about future uses of this, for example, we might want to only round to the nearest whole icon, or we might want to make these interactive at some point. For now though, I think this works for the required use case.

_Requested changes:_
I would suggest adding some logic tests to make sure we are rounding the values correctly. The Playroom is also not working correctly as it cannot import the icons.

maggiedbaker reviewed (approved):
looks good! I also like the idea of defaulting to a star icon, but either way is fine. Just curious, do we have any use cases currently that wouldn't be a star?

danidewitt commented:
@noranda and @maggiedbaker  most of your questions are answered in the [thread](https://github.com/ezcater/recipe/pull/763#discussion_r814007308) above.

- adding default icon: this is definitely a performance issue, but also potentially a licensing issue.
- cases that aren't a star: yes, we have price ratings implemented several different ways in other projects, and i can imagine having heart rating or other icon ratings
- rounding to a whole number: if you want to round to something less precise than 0.5, i think this is fine to happen when you use the icon, so `Math.floor(value)` for your `value` prop (or `Math.ceil()`, you can decide if you want to be less precise)
- making it interactive: we can always add another interactive implementation of ratings later (maybe it's even a separate component that uses this read-only component)

@noranda i added tests and fixed the playroom

danidewitt commented:
visual tests passing
<img width="262" alt="Screen Shot 2022-02-24 at 4 29 48 PM" src="https://user-images.githubusercontent.com/3790037/155610921-b28cf0ba-75b1-4585-987e-8ceec088d02a.png">

noranda reviewed (changes requested):
The logic tests and playroom look great!

One last suggestion:

We should probably have boundaries for the values that extend outside the `0 - max` range. Currently a value of 100 will render 100 stars. We could restrict the values just in case (if value > max, value = max & if value < 0, value = 0).

danidewitt commented:
@noranda good catch - i added a check for value > max and value < 0 and tests for both of those cases.

danidewitt commented:
visual tests passing 
<img width="262" alt="Screen Shot 2022-02-25 at 11 16 34 AM" src="https://user-images.githubusercontent.com/3790037/155749490-2d5354b0-fda6-444f-8d13-84a7ea69562c.png">

noranda reviewed (approved):
Good to go!

danidewitt commented:
tests passing
<img width="366" alt="Screen Shot 2022-02-25 at 6 17 52 PM" src="https://user-images.githubusercontent.com/3790037/155815817-b7d50452-fb33-415f-9dd3-110233764b7c.png">
